### ISSUE TITLE
perf(bitnet/cpu_x86): I2_S decode optimization plan — lm_head sampling sketch, GEMV bandwidth efficiency, <2-bpw packing, ReLU² sparsity

### ISSUE THREAD

**Issue by geisten** [geisten/geistlib#102]
## Context

The BitNet b1.58 2B-4T implementation (I2_S, `cpu_x86` AVX-512-VNNI) is already state of the art among *portable* engines: biased-u8 VPDPBUSD (`dot = Σcode·a − Σa`), x4 row-interleave repack (one activation load feeds 4 rows), spec-head for the F16 lm_head. It beats bitnet.cpp on the 9950X, and the comparison against the literature (T-SAR, FairyFuse) is in `benchmark/TERNARY_BITNET.md`.

**Fresh baseline (9950X, 16T active, `ggml-model-i2_s.gguf`, mean-of-5, current `main`):**

| Metric | Value |
|---|---|
| prefill pp128 | **1037 t/s** |
| decode tg128 | **91.7 t/s** (≈10.9 ms/token) |

`GEIST_PROFILE_FORWARD` breakdown (decode-dominated): FFN ≈ ⅔ of layer time (gate_up 63 %, down 32 %), attention ⅓ (qkv 46 %, o_proj 28 %), lm_head ≈ 3.4 ms/token (Q8 path, sampling). The ternary layer GEMVs read ~520 MB/token and run at **~52–60 GB/s of a ~78 GB/s practical ceiling** → decode is bandwidth-bound with ~25–35 % headroom left; the lm_head is ~30 % of token time.

Research check (papers 2024–2026 + community): our VPDPBUSD multiply design sits on the **right** side of the LUT-vs-multiply crossover on Zen 5 (the bitnet.cpp paper, arXiv:2502.11880, measures the TBL chain slower than MAD; compilade removed the LUT from TQ1_0, ggml PR #8151; FairyFuse arXiv:2604.20913 confirms L1 pressure from LUTs). The remaining levers are **fewer bytes per weight, exploiting free ReLU² sparsity, shrinking the lm_head share, and GEMV bandwidth efficiency**.

## Plan — phases, each with a measurement gate

Working rule (as in `benchmark/TERNARY_BITNET.md`): measure the baseline cold → make one change → measure again → **keep it only if it wins**; record the result in `benchmark/BENCHMARK_X86.md`. Reference command:

```sh
GEIST_BENCH_BACKEND=cpu_x86 OMP_NUM_THREADS=16 OMP_WAIT_POLICY=active \
  bin/linux/release/tests/bench_perf_sweep --gguf ~/models/bitnet/ggml-model-i2_s.gguf \
  --seq-lens 128 --decode-n 128 --repeats 5
```

### Phase 0 — measurability and hygiene (small, immediate)
- [ ] `bench_perf_sweep`: change the default backend from `cpu_neon` (plus a silent `cpu_scalar` fallback!) to `"auto"` — today, benchmarking on x86 without `GEIST_BENCH_BACKEND` accidentally measures the scalar path (4.8 t/s instead of 1037).
- [ ] Document a decode thread sweep (4/8/12/16/24/32 T, `OMP_WAIT_POLICY=active`): bandwidth-bound decode may saturate below 16 T; the best value becomes the new reference. Check whether the CCD-aware threading idea from `hw_probe` (n_l3_domains, "decode pins to one L3 domain") applies to the i2s GEMVs at all — today they are plain `omp parallel for`.
- [ ] THP for the weight arena: `madvise(MADV_HUGEPAGE)` on the mapped GGUF plus the x4/aux blobs (1.1 GB of weights → TLB relief in the streaming GEMV; community reports ggml #12444/#2251). Expectation: low single-digit %; nearly free.

### Phase 1 — shrink the lm_head share (~30 % of token time)
- [ ] **Spec-head for sampling**: today the i8 sketch (82 MB instead of 329/656 MB) only kicks in for greedy. For top-k/top-p sampling: sketch → top-K candidates → exact f16 logits for those only → sample over the candidate set. For top-k sampling with k ≤ TOP_K this is distribution-identical as long as the candidate set contains the true top-k (same recall criterion as the greedy case; document a `GEIST_SPEC_TOPK` sweep). Expectation: lm_head 3.4 → ~1.5 ms ⇒ **+15–20 % decode** on the sampling path.
- [ ] Evaluate **Q8 rescoring instead of F16 rescoring**: compute finalist dots from the existing Q8 blob (half the bytes) rather than f16 — adopt only if greedy output stays byte-identical (otherwise discard; cosine 0.99999 is not necessarily enough for greedy).
- [ ] Optional (measurement gate): a frequency-sorted vocabulary shortlist as a phase-0 filter ahead of the sketch (FR-spec idea, ggml #25187) — only if the sketch read (82 MB) still dominates afterwards.

### Phase 2 — move the decode GEMVs closer to the bandwidth ceiling (largest block: ~520 MB/token)
- [ ] **Fuse per layer stage rather than per pair**: `GEIST_I2S_PAIR` (gate+up) was bandwidth-neutral, but the profile shows qkv as 3 separate OMP regions with 3× activation quantization of the same vector. One region for q+k+v (one quantization, one fork/join) and one for gate+up; vectorize the scalar `quantize_act_row` with AVX2 while at it (absmax + round). Expectation: 2–5 % decode, less variance at low thread counts.
- [ ] Test **`prefetchnta` / non-temporal hints** on the weight stream (keep L3 free for KV and activations). On the A76 software prefetch was a no-op — measure separately on Zen 5, expectation unclear.
- [ ] **First-touch/thread pinning of the x4 blobs**: perform the x4 repack in parallel, in thread slabs, so each thread first-touches "its" rows on its own CCD (ggml #12759 shows a repack without placement can *degrade* TG).

### Phase 3 — bytes per weight below 2.0 (the structural lever)
- [ ] **TQ1_0-style 1.6875-bpw packing** (5 trits per byte, 3⁵=243<256) as a load-time repack variant of the x4 blob: unpack via pow3 multiply + top bits (compilade, ggml PR #8151 — explicitly *without* a LUT). −15.6 % weight bytes ⇒ theoretically +12–15 % decode; Zen 5 has the VNNI compute slack for it. **Mind the measurement gate**: ik_llama.cpp found IQ1_BN (1.625 bpw) sometimes *slower* than 2.0 bpw — unpack cost can eat the bandwidth gain; keep only on a measured win. Correctness: the existing scalar oracle (`test_i2s_gemv_unit` pattern, Δ=0 against `i2s_gemv_m1_scalar`).
- [ ] **ReLU² sparsity in down_proj**: BitNet 2B-4T uses squared ReLU deliberately for sparsity (arXiv:2504.12285; a4.8 measures ~45 % zeros). After `relu²(gate)·up` the zeros in the 6912-wide intermediate vector are **known for free** (no predictor needed). Step 1 (cheap, decides everything): instrument a histogram of all-zero 64-blocks per token. Step 2 only if block sparsity ≥ ~40 %: column-blocked down_proj repack + block skip in the GEMV. Potential: 15–25 % less layer traffic.

### Phase 4 — prefill (already strong; fine-tuning only)
- [ ] Token-tile sweep in the x4 GEMM (`I2S_X4_TT` 2 → 4/8) plus `I2S_JT` in the packed GEMM; bitnet.cpp gained 1.15–2.1× from tile tuning (January 2026 update). A constant sweep only, no restructuring.
- [ ] Add a reference comparison against ik_llama.cpp IQ2_BN on the same host to `BENCHMARK_X86.md` (sanity bar; 7950X reference: PP512 ≈ 520–540 t/s at 3B).

### Deliberately NOT pursued (research outcome, rationale documented in the comments)
- LUT kernels (T-MAC/TL1/TL2) on Zen 5 — multiply beats LUT on wide AVX-512 cores (arXiv:2502.11880, ggml #8151).
- FairyFuse-style multiply-free FP masking — loses against VNNI (arXiv:2604.20913).
- AMX-INT8 — no AMX on AMD; only on a concrete Xeon requirement.
- BitNet a4.8/v2 INT4 activations — needs a differently trained checkpoint; W1.58A8 stays. Per-block activation scales likewise (deviates from training, accuracy risk without a bandwidth gain).
- In-register LUT (T-SAR) — hardware co-design, not available on stock silicon.

## Acceptance criteria

- [ ] Every adopted phase change: measurement (mean-of-5, cold, protocol as in `benchmark/BENCHMARK_X86.md`) documented in the PR; regressions are discarded.
- [ ] Correctness per kernel change: Δ=0 against the scalar oracle (extend the existing `test_i2s_*` unit patterns); the spec-head's greedy output stays byte-identical to the exact head; new formats and paths also run under `GEIST_FORCE_ISA=avx2` (CI gate from #96) without SIGILL or errors.
- [ ] `make test-unit test-int test-e2e` green (the x86 CI track, including the `GEIST_FORCE_ISA=avx2` run).
- [ ] Target after phases 1+2: **decode ≥ 105 t/s** on the 9950X reference setup (+15 %); after phase 3 (if the measurement gates hold): **≥ 115 t/s**. Prefill ≥ 1037 t/s (no regression).
- [ ] `benchmark/BENCHMARK_X86.md` and `benchmark/TERNARY_BITNET.md` updated (the new baseline 1037/91.7 replaces 884/77.9).

**Comment by geisten** [comment]
Phasen 0–4 umgesetzt und vermessen in PR #103 — Kurzfassung:

**Gewinne** (9950X, mean-of-5): decode 91.7 → **94.9** (+3.5 %), prefill 1036.7 → **1093** (+5.4 %). Komponenten: THP (+1.3 %), Pair-Fusion default ON (+1.2 %), NTA-Prefetch 512 B (+2.7 % A/B), Prefill-Tile TT=4 (+5 % prefill). Gemma-Regression: keine (A/B 498/45.6 → 501/45.5).

**Bit-Exaktheits-Funde**: (1) `-freciprocal-math` machte die Pair-Fusion 1 ULP untreu (noinline-Fix + verschärfter byte-exakter Unit-Test mit distinct weights); (2) Spec-Head-Phase-3 war nur trajektorien-glücklich byte-identisch — jetzt per Konstruktion (gleiche Kernel-Instanz via Ein-Zeilen-View); (3) SPEC_TOPK default 512→1024 (Recall-Marge).

**Messgates, die NICHT bestanden** (dokumentiert, nicht geshippt):
- ❌ **Spec-Head fürs Top-k-Sampling** (Phase 1, war auf +15–20 % geschätzt): Stride-4-Sketch-Ranking jenseits Rank 1 ist zu verrauscht — echte Top-40 lagen außerhalb von Top-8192-Kandidaten; exakter Recall ⇒ Stride 1 ⇒ gleiche Bytes wie der Q8-Dense-Head ⇒ kein Gewinn. Regressionstest pinnt den Dense-Fallback.
- ❌ **ReLU²-Block-Skip** (Phase 3): 58.2 % Element-Sparsity, aber **0.00 %** All-Zero-64er-Blöcke über alle 30 Layer (Probe-Erweiterung in `GEIST_DUMP_ACT_SPARSITY`). Unstrukturierte Sparsity ist im Streaming-GEMV bei 0.25 B/wt nicht verwertbar.
- ❌ TT=8 (Register-Spills), Act-Quant-Vektorisierung (war schon auto-vektorisiert), First-Touch (war schon korrekt).

**Bleibt offen** (dieses Issue): TQ1_0-artiges **<2-bpw-Packing** — der letzte strukturelle Decode-Hebel (−15.6 % Weight-Bytes). Braucht ein eigenes Format-Redesign (Base-3 mit 5 Trits/Byte ist mit dem x4-Row-Interleave-Byte-Layout unvereinbar): neuer Load-Time-Repack + pow3-Multiply-Unpack im VPDPBUSD-Loop, Messgate wie gehabt (ik_llama.cpp-Warnung: IQ1_BN war teils langsamer als 2.0 bpw).

Die Ziele „decode ≥105 nach Phase 1+2" hingen am Sampling-Spec-Head, der sein Exaktheits-Gate nicht bestand — die ehrliche Ausbeute ist +3.5 % decode / +5.4 % prefill plus die Korrektheits-Härtung.

**Comment by geisten** [comment]
Die offenen Punkte sind in **#104** verschoben (TQ1_0-artiges 1.6-bpw-Decode-Packing als eigenständiges Format-/Kernel-Projekt mit Feasibility-Kill-Switch, plus der ik_llama.cpp-IQ2_BN-Referenzlauf). Dieses Issue kann mit dem Merge von PR #103 geschlossen werden — alle Phasen 0–4 sind umgesetzt oder mit Messwerten als Dead-End dokumentiert.

**Comment by geisten** [comment]
**Statusprüfung gegen den Code** — mehrere Punkte sind erledigt, die Checkboxen aber noch offen. Ohne 9950X kann ich nichts messen, deshalb hier nur, was im Code nachweisbar ist.

### Phase 0

- ✅ **`bench_perf_sweep`-Default auf `"auto"`** — behoben. `tests/bench_perf_sweep.c:169` trägt den Kommentar samt Verweis: *„The old default of cpu_neon + silent cpu_scalar fallback measured the scalar path on x86 builds unless GEIST_BENCH_BACKEND was set (#102)."* Das war der schlimmste Punkt der Liste: ein Benchmark, der still den falschen Pfad misst (4,8 statt 1037 t/s).
- ✅ **THP für die Weight-Arena** — `src/io/gguf_reader.c:473` ruft `madvise(MADV_HUGEPAGE)` auf der Mapping-Region, plus `MADV_WILLNEED`.
- ❌ **Thread-Sweep dokumentiert** — offen. `benchmark/BENCHMARK_X86.md` existiert nicht; die Referenz für den Sweep fehlt also noch.

### Phase 1

- ✅ **Spec-Head fürs Sampling** — *versucht und wegmessen*, nicht offen. `src/archs/transformer/forward/spec_head.c:635` hält das Ergebnis fest: der stride-4-Sketch rankt jenseits von Rang 1 zu verrauscht — auf BitNet 2B-4T landen echte Top-40-Zeilen außerhalb der groben Top-8192. Perfekte Recall hätte stride-1 gebraucht, also dieselbe Byte-Zahl wie der dichte Q8-Head: kein Gewinn.

  Das Bemerkenswerte ist, dass das Negativergebnis **festgenagelt** ist: `tests/test_spec_head_sampling_int.c` prüft, dass der Spec-Head bei `temperature > 0` gar nicht erst anspringt. Ein zukünftiger Versuch, ihn dort zu aktivieren, wird rot — die gemessene Erkenntnis kann nicht stillschweigend verlorengehen.

### Was ich nicht beurteilen kann

Alles ab Phase 2 ist messungsgesteuert („measure the baseline cold → make one change → measure again → keep it only if it wins") und braucht den 9950X. Ohne die Maschine wäre jede Änderung dort eine Behauptung, kein Ergebnis — und dieses Ticket ist ausdrücklich anders angelegt.

**Vorschlag:** die erledigten Checkboxen abhaken, damit die verbleibende Arbeit sichtbar wird. Der einzige Phase-0-Rest — der Thread-Sweep und `BENCHMARK_X86.md` — ist die natürliche nächste Einheit, sobald jemand an der Hardware sitzt.

**Comment by geisten** [comment]
Aufgeteilt und geschlossen. Der Rest lebt in **#212** weiter, so wie #104 schon den t5-Teil übernommen hat.

## Was dieser Plan erreicht hat

9950X, 16 T `OMP_WAIT_POLICY=active`, mean-of-5 — Ledger in `benchmark/results/X86.md`:

| Metrik | Start | Ende | Δ |
| :-- | --: | --: | --: |
| prefill pp128 | 1036,7 | **1098** | +5,9 % |
| decode tg128 | 91,7 | **103,1** | +12,4 % |

Das Prefill-Kriterium („kein Regress unter 1037") ist erfüllt. Das Decode-Ziel „≥ 105 nach Phase 1+2" wurde mit 103,1 **knapp verfehlt**, „≥ 115 nach Phase 3" nicht erreicht — beides steht so im Ledger, nicht schöngerechnet.

## Übernommen (gemessen, behalten)

| Phase | Änderung | Wirkung |
| :-- | :-- | :-- |
| 0 | `bench_perf_sweep` Default-Backend `"auto"` | maß auf x86 vorher versehentlich den Skalarpfad (4,8 statt 1037 t/s) |
| 0 | THP auf >2 MB-Heap-Blobs + GGUF-mmap | 91,7 → 92,9 decode |
| 0 | Thread-Sweep dokumentiert | 16 T optimal; SMT (32 T) kostet −24 % decode |
| 2 | `GEIST_I2S_PAIR` default ON (q+k / gate+up fusioniert) | 93,4 → 94,6 (+1,2 %) |
| 2 | `prefetchnta` auf dem x4-Gewichtsstrom, 512 B voraus | +2,7 % (A/B 94,5 vs 92,0); 256/512/1024 gesweept |
| 4 | Prefill-Token-Tile `I2S_X4_TT` 2 → 4 | 1093 prefill |
| 3 (via #104) | t5 base-3-Packing, 1,6 bpw, pow3-Unpack | **103,1 decode** (+10,5 % A/B vs x4) |

## Gemessen und verworfen — das Wertvollste hier

Drei Hypothesen wurden ausprobiert und **widerlegt**. Sie stehen im Ledger unter „Measured dead ends", damit sie niemand erneut anfasst:

- **Spec-Head für top-k-Sampling** (Phase 1, der größte erhoffte Einzelgewinn): das Rangrauschen des stride-4-Sketch jenseits von Rang 1 ist enorm — auf BitNet 2B-4T landen echte top-40-Zeilen außerhalb sogar der top-8192-Grobkandidaten. Exakte Recall bräuchte eine volle Breite in Phase 1, also genau die Bytes des Q8-Dense-Head: kein Gewinn. Festgehalten in `spec_head.c` und `tests/test_spec_head_sampling_int.c`.
- **ReLU²-Block-Skip im down_proj** (Phase 3): der Instrumentierungs-Gate wurde gebaut (`forward/probes.c`) und hat die Idee erledigt — **58 % Element-Sparsity, aber 0,00 % vollständig leere 64er-Blöcke.** Die Nullen clustern nicht, es gibt nichts zu überspringen. Genau dafür war Schritt 1 als Gate vor Schritt 2 gedacht.
- **`I2S_X4_TT=8`**: Registerspills, 811 t/s.

## Offen → #212

Threading/Platzierung (CCD-bewusste Verteilung, First-touch der x4/t5-Blobs), Q8-Rescoring der lm_head-Finalisten, Vektorisierung von `quantize_act_row`, der nie durchgeführte `I2S_JT`-Sweep und die ik_llama-Referenzzeile.

Der Grund für die Trennung: was hier offen ist, sind fünf konkrete, unabhängige Hebel — in einem Plan mit dreißig abgehakten Punkten und drei widerlegten Hypothesen sieht man sie nicht mehr.
